Maxime/dogstatsd add set

hush-hush opened the pull request:
### What does this PR do?

#### First:

The number of metrics continues to increase. This commit split metrics
one per file.

Renaming 'Sampler' struct to TimeSampler (ie the sampler for
dogstatsd). This make the documentation clearer as we now have a "sampler"
concept implemented by two class: TimeSampler and CheckSampler.

Updated the README to add a diagram of the aggregator workflow.

Rename 'Metrics' to 'ContextMetrics' to avoid confusion between set of
metrics and metrics ordered by context (since this is used by both
sampler).

#### Second:
Propagate the MetricSample down to the each metric. Every metric is now
in charge of it's internal computation.

This will allow metrics to be based on something else than float64 (as
needed by Set). MetricSample will be extended to contain other kind of
sample.

hush-hush committed:
Split and rename internal struct and files to prepare future development
of the aggregator.

The number of metrics continues to increase. This commit split metrics
one per file.

Renaming 'Sampler' struct to TimeSampler (ie the sampler for
dogstatsd). This make the documentation clearer as we now have a "sampler"
concept implemented by two class: TimeSampler and CheckSampler.

Updated the README to add a diagram of the aggregator workflow.

Rename 'Metrics' to 'ContextMetrics' to avoid confusion between set of
metrics and metrics ordered by context (since this is used by both
sampler).

hush-hush committed:
Propagate the MetricSample down to the each metric. Every metric is now
in charge of it's internal computation.

This will allow metric to be based on something else than float64 (as
needed by Set). MetricSample will be extended to contain other kind of
sample.

olivielpeau reviewed (commented):
Thanks for moving these things around, looks more organized now :) The current changes look good to me 👍 

There's just a typo in the file names: `monotic_count` -> `monotonic_count` (same thing for the test file)

hush-hush committed:
Add Set to dogstatsd

Set counts the number of unique value for a metric. It takes advantage
from the new RawValue.

For now RawValue is also set for other kind of metrics even if we will
most likeley never use it.

hush-hush committed:
Clean parser tests

- Removed empty test related to point tested elsewhere
- use require instead of condition

olivielpeau reviewed (approved):
A couple of comments, but other than that LGTM! 👌   (also, big 👍 for the commits and their descriptions)

Feel free to merge once you've addressed them
